mountpoint-s3-fs/src/upload/incremental.rs

```diff
@@ -234,8 +247,24 @@ where
     }
 
     pub async fn get_buffer(&mut self, capacity: usize) -> Result<UploadBuffer, UploadError<Client::ClientError>> {
-        let checksum_algorithm = self.checksum_algorithm.get_mut().await?.unwrap().clone();
-
+        let checksum_algorithm = if let Some(checksum_algorithm) = self.checksum_algorithm.clone() {
+            checksum_algorithm
+        } else {
+            match self
+                .event_receiver
+                .recv()
+                .await
+                .map_err(|_| UploadError::UploadAlreadyTerminated)?
+            {
+                AppendUploadEvent::ChecksumAlgorithm(checksum_algorithm) => {
+                    self.checksum_algorithm.insert(checksum_algorithm).clone()
+                }
+                AppendUploadEvent::PutResponse(_) => {
+                    unreachable!("cannot receive a response event before the checksum algorithm");
+                }
+                AppendUploadEvent::Error(upload_error) => return Err(upload_error),
+            }
+        };
         while self.requests_in_queue > 0 {
             match UploadBuffer::try_new(capacity, &checksum_algorithm, &self.pool, self.mem_limiter.clone())? {
```

Comment:
not related: is there a limit for how much a single "append" file handle can reserve for its buffers? the total amount of free RAM is the only limit?